File: src/main/java/practice/account/domain/Account.java
package practice.account.domain;

import java.math.BigDecimal;
import java.util.ArrayList;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.experimental.FieldDefaults;

@Getter
@Builder
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Account {

  Long accountId;
  Long userId;
  String accountNumber;

  AccountType type;
  BigDecimal balance;

  static final BigDecimal DAILY_WITHDRAWAL_LIMIT = BigDecimal.valueOf(3_000_000);
  BigDecimal todayWithdraw;

  @Builder.Default
  TransactionHistoryList transactionHistoryList = new TransactionHistoryList(new ArrayList<>());

  public boolean canNotWithdrawNow(BigDecimal amount) {
    return !(AccountType.MAIN.equals(type)
        && balance.compareTo(amount) >= 0);
  }

  public boolean isExternalWithdrawLimitExceeded() {
    return !(AccountType.MAIN.equals(type)
        && todayWithdraw.compareTo(DAILY_WITHDRAWAL_LIMIT) < 0);
  }

  public BigDecimal calculateRequiredAmount(BigDecimal amount){
    return amount.subtract(balance);
  }

  public void withdraw(BigDecimal amount, Account targetAccount) {
    if(canNotWithdrawNow(amount)) {
      return;
    }

    balance = balance.subtract(amount);
    transactionHistoryList.add(TransactionType.WITHDRAW, amount, new AccountInfo(targetAccount));
  }

  public void withdraw(BigDecimal amount, Long paymentId) {
    if(canNotWithdrawNow(amount)) {
      return;
    }

    balance = balance.subtract(amount);
    transactionHistoryList.add(TransactionType.PAYMENT, amount, paymentId);
  }

  public void deposit(BigDecimal amount, Account sourceAccount) {
    balance = balance.add(amount);
    transactionHistoryList.add(TransactionType.DEPOSIT, amount, new AccountInfo(sourceAccount));
  }

  public void deposit(TransactionHistory transactionHistory) {
    balance = balance.add(transactionHistory.getAmount());
    transactionHistory.setToDeposit();
    transactionHistoryList.add(transactionHistory);
  }




}

File: src/main/kotlin/practice/account/domain/Account.kt
package practice.account.domain

import java.math.BigDecimal

class Account(
    var accountId: Long? = null,
    var userId: Long? = null,
    var accountNumber: String? = null,
    var type: AccountType? = null,
    var balance: BigDecimal? = null,
    var todayWithdraw: BigDecimal? = null,
    var transactionHistoryList: TransactionHistoryList = TransactionHistoryList()
) {

    companion object {
        private val DAILY_WITHDRAWAL_LIMIT = BigDecimal.valueOf(3000000)
    }

    fun canNotWithdrawNow(amount: BigDecimal?): Boolean {
        return !(AccountType.MAIN == type && balance!! >= amount)
    }

    fun isExternalWithdrawLimitExceeded(): Boolean {
        return !(AccountType.MAIN == type && todayWithdraw!! < DAILY_WITHDRAWAL_LIMIT)
    }

    fun calculateRequiredAmount(amount: BigDecimal): BigDecimal {
        return amount.subtract(balance)
    }

    fun withdraw(amount: BigDecimal, targetAccount: Account) {
        if (canNotWithdrawNow(amount)) {
            return
        }

        balance = balance!!.subtract(amount)
        transactionHistoryList.add(TransactionType.WITHDRAW, amount, AccountInfo(targetAccount))
    }

    fun withdraw(amount: BigDecimal, paymentId: Long) {
        if (canNotWithdrawNow(amount)) {
            return
        }

        balance = balance!!.subtract(amount)
        transactionHistoryList.add(TransactionType.PAYMENT, amount, paymentId)
    }

    fun deposit(amount: BigDecimal, sourceAccount: Account) {
        balance = balance!!.add(amount)
        transactionHistoryList.add(TransactionType.DEPOSIT, amount, AccountInfo(sourceAccount))
    }

    fun deposit(transactionHistory: TransactionHistory) {
        balance = balance!!.add(transactionHistory.amount)
        transactionHistory.setToDeposit()
        transactionHistoryList.add(transactionHistory)
    }
}